File: go/closure/polymorphism/mammal.go
package polymorphism

import "fmt"

type Mammal struct {
	speak     func()
	getWeight func() int
	setWeight func(int)
}

func (m Mammal) print() {
	fmt.Println("Printing...")
	fmt.Print("\t")
	m.speak()
	fmt.Println("\tWeight: ", m.getWeight())
}

func Cow() Mammal {
	weight := 100

	speak := func() { fmt.Println("momo~~ momo~~") }
	getWeight := func() int { return weight }
	setWeight := func(w int) { weight = w }

	return Mammal{speak, getWeight, setWeight}
}

func Sheep() Mammal {
	weight := 50

	speak := func() { fmt.Println("mee~~ mee~~") }
	getWeight := func() int { return weight }
	setWeight := func(w int) { weight = w }

	return Mammal{speak, getWeight, setWeight}
}

File: java/src/closure/polymorphism/Mammal.java
package closure.polymorphism;

public interface Mammal {
    void speak();

    int getWeight();

    void setWeight(int w);

    default void print() {
        System.out.println("Printing...");
        System.out.print('\t');
        speak();
        System.out.printf("\tWeight: %d\n", getWeight());
    }
}

class Main {
    static Mammal Cow() {
        int cowWeight = 100;

        return new Mammal() {
            int weight = cowWeight;

            @Override
            public void speak() {
                System.out.println("momo~~ momo~~");
            }

            @Override
            public int getWeight() {
                return weight;
            }

            @Override
            public void setWeight(int w) {
                weight = w;
            }
        };
    }

    static Mammal Sheep() {
        int sheepWeight = 50;

        return new Mammal() {
            int weight = sheepWeight;

            @Override
            public void speak() {
                System.out.println("mee~~ mee~~");
            }

            @Override
            public int getWeight() {
                return weight;
            }

            @Override
            public void setWeight(int w) {
                weight = w;
            }
        };
    }
}
